streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -924,28 +924,69 @@ private StreamThread createStreamThread(final long cacheSizePerThread, final int
      * @return name of the added stream thread or empty if a new stream thread could not be added
      */
     public Optional<String> addStreamThread() {
-        synchronized (changeThreadCount) {
```

Comment:
I moved the changeThreadCount lock to be more specific so we will not hold a lock while acquiring a new one.

Reply:
Nice. One minor suggestion, log a warning if the client isn't running (or rebalancing) and print the current state